### PR TITLE
feat(telemetry): telemetry metrics support

### PR DESCRIPTION
**Pull request by avara1986**
# Description
Split telemetry worker: One Worker to report events (heartbeat, startup...) Other to report metrics.
Class methods to report distributions, gauge, count and rate Metrics type. 

This PR is part of the main PR https://github.com/DataDog/dd-trace-py/pull/5046

# Motivation
ASM needs to report information about 

# PRs related:
System tests: 
- https://github.com/DataDog/system-tests/pull/823
- https://github.com/DataDog/system-tests/pull/890

This PR comes from
- https://github.com/DataDog/dd-trace-py/pull/5046
- https://github.com/DataDog/dd-trace-py/pull/4943 cc @mabdinur 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer is aware of, and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Comment by avara1986** [comment]
CAVEAT: internal folder coverage isn't working correctly :face_with_spiral_eyes:

File: ddtrace/internal/writer.py [hunk 1]
```diff
@@ -250,7 +251,8 @@ def __init__(
         max_payload_size=None,  # type: Optional[int]
         timeout=agent.get_trace_agent_timeout(),  # type: float
         dogstatsd=None,  # type: Optional[DogStatsd]
-        report_metrics=False,  # type: bool
+        report_health_metrics=False,  # type: bool
+        report_telemetry_metrics=True,  # type: bool
```
Comment (mabdinur) [inline ddtrace/internal/writer.py:255]:
```suggestion
```

File: ddtrace/internal/writer.py [hunk 2]
```diff
@@ -326,7 +328,8 @@ def __init__(
         if additional_header_str is not None:
             self._headers.update(parse_tags_str(additional_header_str))
         self.dogstatsd = dogstatsd
-        self._report_metrics = report_metrics
+        self._report_health_metrics = report_health_metrics
+        self._report_telemetry_metrics = report_telemetry_metrics
```
Comment (mabdinur) [inline ddtrace/internal/writer.py:332]:
```suggestion
```

File: ddtrace/settings/config.py [hunk 3]
```diff
@@ -222,6 +222,8 @@ def __init__(self):
 
         self.health_metrics_enabled = asbool(os.getenv("DD_TRACE_HEALTH_METRICS_ENABLED", default=False))
 
+        self.telemetry_metrics_enabled = asbool(os.getenv("DD_TELEMETRY_METRICS_ENABLED", default=False))
```
Comment (mabdinur) [inline ddtrace/settings/config.py:225]: We can keep this setting internal. This way we can easily remove it if it’s rarely used. 

```suggestion
        self.telemetry_metrics_enabled = asbool(os.getenv("_DD_TELEMETRY_METRICS_ENABLED", default=False))
```

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -460,7 +461,8 @@ def configure(
                 sampler=self._sampler,
                 priority_sampler=self._priority_sampler,
                 dogstatsd=get_dogstatsd_client(self._dogstatsd_url),
-                report_metrics=config.health_metrics_enabled,
+                report_health_metrics=config.health_metrics_enabled,
+                report_telemetry_metrics=config.telemetry_metrics_enabled,
```
Comment (mabdinur) [inline ddtrace/tracer.py:465]:
```suggestion
```

File: ddtrace/tracer.py [hunk 5]
```diff
@@ -240,7 +240,8 @@ def __init__(
...
```
Comment (mabdinur) [inline ddtrace/tracer.py:244]: We should avoid changing the init args of the AgentWriter if possible. 
```suggestion
```

File: docs/configuration.rst [hunk 6]
```diff
@@ -66,6 +66,12 @@ The following environment variables for the tracer are supported:
      description: |
          Enables sending :ref:`telemetry <Instrumentation Telemetry>` events to the agent.
 
+   DD_TELEMETRY_METRICS_ENABLED:
```
Comment (mabdinur) [inline docs/configuration.rst:69]: We don’t need to add this new configuration to our public docs. We can use it internally for tests. 
```suggestion
```

File: ddtrace/internal/telemetry/writer.py [hunk 7]
```diff
@@ -70,50 +88,68 @@ def send_event(self, request):
 
     def get_headers(self, request):
         # type: (Dict) -> Dict
-        """Get all telemetry api v1 request headers"""
+        """Get all telemetry api v2 request headers"""
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:91]: Telemetry api v2 is not necessarily backwards compatible with v1. We have to make sure we’re sending all the new mandatory fields defined in the instrumentation telemetry spec.

File: ddtrace/internal/telemetry/writer.py [hunk 8]
```diff
@@ -123,11 +159,151 @@ def _flush_events_queue(self):
...
+
+    def __init__(self):
+        # type: () -> None
+        super(TelemetryMetricsWriter, self).__init__(interval=_get_telemetry_metrics_interval_or_default())
+        self._namespace = MetricNamespace()
+
+    def _flush_namespace_metrics(self):
+        # type () -> List[Metric]
+        """Returns a list of all generated metrics and clears the namespace's list"""
+        log.debug("collect Telemetry Metrics")
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:199]: Can we add this log line after we flush the metrics queue (and outside of the `with lock` block). We can then add some more useful info to this log message like the number of metrics flushed.

File: ddtrace/internal/telemetry/writer.py [hunk 9]
```diff
@@ -123,11 +159,151 @@ def _flush_events_queue(self):
...
+        # type: (NamespaceMetricType) -> None
+        for payload_type, namespaces in namespace_metrics.items():
+            for namespace, metrics in namespaces.items():
+                if metrics:
+                    payload = {
+                        "namespace": namespace,
+                        "lib_language": "python",
+                        "lib_version": _pep440_to_semver(),
+                        "series": [m.to_dict() for m in metrics.values()],
+                    }
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:219]: Random thought (feel free to ignore); We can add more detail to this log message. Also what’s the expect size of the payloads. Could this significantly increase the size of a users debug logs? Could we print a few fields instead of the full payload?

File: ddtrace/internal/telemetry/writer.py [hunk 10]
```diff
@@ -123,11 +159,151 @@ def _flush_events_queue(self):
...
+            self._client.send_event(telemetry_event)
+
+    def on_shutdown(self):
+        self.periodic()
+
+
+class TelemetryWriter(TelemetryBase):
+    """
+    Submits Instrumentation Telemetry events to the datadog agent.
+    Supports v1 of the instrumentation telemetry api
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:281]: ?
```suggestion
    Supports v2 of the instrumentation telemetry api
```

File: ddtrace/internal/telemetry/writer.py [hunk 11]
```diff
@@ -70,50 +88,68 @@ def send_event(self, request):
 
     def get_headers(self, request):
         # type: (Dict) -> Dict
-        """Get all telemetry api v1 request headers"""
+        """Get all telemetry api v2 request headers"""
         headers = self._headers.copy()
-        headers["DD-Telemetry-Debug-Enabled"] = request["debug"]
+        headers["DD-Telemetry-Debug-Enabled"] = str(request["debug"]).lower()
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:93]: The debug field should not be lower case ([PR](https://github.com/DataDog/dd-trace-py/commit/29826f959af174d44fd5af107be3631a0c170763))
```suggestion
        headers["DD-Telemetry-Debug-Enabled"] = request["debug"]
```

File: ddtrace/internal/telemetry/writer.py [hunk 12]
```diff
@@ -19,19 +20,35 @@
...
 
-def _get_interval_or_default():
+def _get_heartbeat_interval_or_default():
     # type: () -> float
     return float(os.getenv("DD_TELEMETRY_HEARTBEAT_INTERVAL", default=60))
 
 
+def _get_telemetry_metrics_interval_or_default():
+    # type: () -> float
+    return float(os.getenv("DD_TELEMETRY_METRICS_POLL_SECONDS", default=10))
```
[the remark this answers is not in the document]
Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py:49]: :+1:

File: docs/configuration.rst (hunk 6 above)
Comment (avara1986) [reply docs/configuration.rst:69]: :+1:

File: ddtrace/internal/telemetry/writer.py [hunk 13]
```diff
@@ -70,50 +88,68 @@ def send_event(self, request):
...
         self._debug = asbool(os.environ.get("DD_TELEMETRY_DEBUG", "false"))
-        forksafe.register(self._fork_writer)
 
         # Counter representing the number of events sent by the writer. Here we are relying on the atomicity
         # of `itertools.count()` which is a CPython implementation detail. The sequence field in telemetry
         # payloads is only used in tests and is not required to process Telemetry events.
         self._sequence = itertools.count(1)
         self._client = _TelemetryClient(self.ENDPOINT_V1)
 
-    def _flush_integrations_queue(self):
```
[the remark this answers is not in the document]
Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py:110]: good point! :clap: :clap: :clap: :clap:

File: ddtrace/internal/telemetry/writer.py [hunk 14]
```diff
@@ -74,46 +92,64 @@ def get_headers(self, request):
...
+class TelemetryBase(PeriodicService):
     """
-    Submits Instrumentation Telemetry events to the datadog agent.
-    Supports v1 of the instrumentation telemetry api
+    Common features of Telemetry services
     """
 
     # telemetry endpoint uses events platform v2 api
     ENDPOINT_V1 = "telemetry/proxy/api/v2/apmtelemetry"
+    # Counter representing the number of events sent by the writer. Here we are relying on the atomicity
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:109]: ENDPOINT_V1 constant should now be ENDPOINT_V2.

File: ddtrace/internal/writer.py [hunk 15]
```diff
@@ -532,6 +536,10 @@ def write(self, spans=None):
                     # are initialized
                     if asbool(os.getenv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", True)):
                         telemetry_writer.enable()
+
+                    if self._report_telemetry_metrics:
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/internal/writer.py:540]: Health metrics are collected and sent in the agent writer while telemetry events should be independent of the agent writer. We enable telemetry collection in the agent writer so that the first telemetry event is sent when the the first trace is sent, but this can change in the future. Ideally we should have one global flag that determines whether telemetry collection is enabled, having two might cause confusion and unexpected behavior down the line. I’d prefer to remove this argument and field from the AgentWriter but if this makes testing significantly more challenging I’m fine with keeping this code as is.

File: ddtrace/internal/writer.py (hunk 1 above)
Comment (mabdinur) [reply ddtrace/internal/writer.py:255]: If we decide to keep this argument the default value should be False. This feature should only be enabled using the globs config.

File: ddtrace/internal/writer.py [hunk 16]
```diff
@@ -250,7 +252,6 @@ def __init__(
         max_payload_size=None,  # type: Optional[int]
         timeout=agent.get_trace_agent_timeout(),  # type: float
         dogstatsd=None,  # type: Optional[DogStatsd]
-        report_metrics=False,  # type: bool
```
Comment (mabdinur) [inline ddtrace/internal/writer.py:253]: We should make this change in another PR. I expect a few tracer tests to fail with this change.

File: ddtrace/internal/writer.py [hunk 17]
```diff
@@ -532,6 +531,10 @@ def write(self, spans=None):
                     # are initialized
                     if asbool(os.getenv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", True)):
                         telemetry_writer.enable()
+
+                    if config.telemetry_metrics_enabled:
```
Comment (mabdinur) [inline ddtrace/internal/writer.py:535]: bug:
```suggestion
                    if config._telemetry_metrics_enabled:
```

File: ddtrace/internal/telemetry/writer.py [hunk 18]
```diff
@@ -74,46 +92,64 @@ def get_headers(self, request):
...
+
+        :param Dict payload: stores a formatted telemetry event
+        :param str payload_type: The payload_type denotes the type of telmetery request.
+            Payload types accepted by telemetry/proxy v1: app-started, app-closing, app-integrations-change
+        """
+        if self._enabled:
+            event = {
+                "tracer_time": int(time.time()),
+                "runtime_id": get_runtime_id(),
+                "api_version": "v1",
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:144]:
```suggestion
                "api_version": "v2",
```

File: ddtrace/internal/telemetry/writer.py [hunk 19]
```diff
@@ -123,11 +159,153 @@ def _flush_events_queue(self):
...
+        self._sequence = itertools.count(1)
+
+    def _stop_service(self, *args, **kwargs):
+        # type: (...) -> None
+        super(TelemetryBase, self)._stop_service(*args, **kwargs)
+        self.join()
+
+    def _fork_writer(self):
+        # type: () -> None
+        pass
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:186]:
```suggestion
       self. _flush_events_queue()
```

File: ddtrace/internal/telemetry/writer.py [hunk 20]
```diff
@@ -123,11 +159,153 @@ def _flush_events_queue(self):
...
+        # type () -> List[Metric]
+        """Returns a list of all generated metrics and clears the namespace's list"""
+        with self._lock:
+            namespace_metrics = self._namespace.get()
+            self._namespace._flush()
+        return namespace_metrics
+
+    def _fork_writer(self):
+        # type: () -> None
+        self._namespace._flush()
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:209]:
```suggestion
        super(TelemetryBase, self)._fork_writer()
        self._namespace._flush()
```

File: ddtrace/internal/telemetry/writer.py (hunk 18 above)
Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py:144]: IIRC, we said in other comment (or other comment in different PR) we should keep v1 at this moment (system tests and other concerns about the migration to v2), right?

File: ddtrace/internal/telemetry/writer.py [hunk 21]
```diff
@@ -123,11 +159,154 @@ def _flush_events_queue(self):
...
+        self._sequence = itertools.count(1)
+
+    def _stop_service(self, *args, **kwargs):
+        # type: (...) -> None
+        super(TelemetryBase, self)._stop_service(*args, **kwargs)
+        self.join()
+
+    def _fork_writer(self):
+        # type: () -> None
+       self. _flush_events_queue()
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:186]:
```suggestion
       self._flush_events_queue()
```

File: ddtrace/internal/telemetry/writer.py (hunk 18 above)
Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py:144]: Yes then we should update the headers

Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py:144]: Nevermind, I was confused by the events platform v2 endpoint

**Review by mabdinur — DISMISSED** [review]

File: ddtrace/internal/telemetry/writer.py [hunk 22]
```diff
@@ -123,11 +159,154 @@ def _flush_events_queue(self):
...
+        self._sequence = itertools.count(1)
+
+    def _stop_service(self, *args, **kwargs):
+        # type: (...) -> None
+        super(TelemetryBase, self)._stop_service(*args, **kwargs)
+        self.join()
+
+    def _fork_writer(self):
+        # type: () -> None
+       self._flush_events_queue()
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:186]:
```suggestion
        # type: () -> None
        self._flush_events_queue()
```

**Review by mabdinur — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
